File: batches_javascript/batch_13_different_implementation_same_algorithm/compare.js
function binarySearchRec(arr, lo, hi, target) {
    if (lo > hi) return -1;
    const mid = lo + Math.floor((hi - lo) / 2);
    if (arr[mid] === target) return mid;
    if (arr[mid] < target) return binarySearchRec(arr, mid + 1, hi, target);
    return binarySearchRec(arr, lo, mid - 1, target);
}

function binarySearch(arr, target) {
    return binarySearchRec(arr, 0, arr.length - 1, target);
}

const a = [1, 3, 5, 7, 9];
console.log(binarySearch(a, 5));
console.log(binarySearch(a, 4));

File: batches_cpp/batch_13_different_implementation_same_algorithm/compare.cpp
#include <iostream>
#include <vector>
using namespace std;

int binarySearchRec(const vector<int>& arr, int lo, int hi, int target) {
    if (lo > hi) return -1;
    int mid = lo + (hi - lo) / 2;
    if (arr[mid] == target) return mid;
    if (arr[mid] < target) return binarySearchRec(arr, mid + 1, hi, target);
    return binarySearchRec(arr, lo, mid - 1, target);
}

int binarySearch(const vector<int>& arr, int target) {
    return binarySearchRec(arr, 0, (int)arr.size() - 1, target);
}

int main() {
    vector<int> a = {1, 3, 5, 7, 9};
    cout << binarySearch(a, 5) << endl;
    cout << binarySearch(a, 4) << endl;
    return 0;
}
